File: src/main/java/com/portofino/realtrainmodunofficial/network/TrainScriptDataPayload.java
package com.portofino.realtrainmodunofficial.network;

import com.portofino.realtrainmodunofficial.RealTrainModUnofficial;
import com.portofino.realtrainmodunofficial.entity.TrainEntity;
import io.netty.buffer.ByteBuf;
import net.minecraft.client.Minecraft;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import net.neoforged.neoforge.network.handling.IPayloadContext;

import java.util.HashMap;
import java.util.Map;

public record TrainScriptDataPayload(int trainEntityId, Map<String, String> data) implements CustomPacketPayload {
    public static final Type<TrainScriptDataPayload> TYPE = new Type<>(
        ResourceLocation.fromNamespaceAndPath(RealTrainModUnofficial.MODID, "train_script_data")
    );

    public static final StreamCodec<ByteBuf, TrainScriptDataPayload> STREAM_CODEC = StreamCodec.of(
        (buf, payload) -> {
            ByteBufCodecs.INT.encode(buf, payload.trainEntityId());
            Map<String, String> map = payload.data();
            ByteBufCodecs.INT.encode(buf, map.size());
            for (Map.Entry<String, String> entry : map.entrySet()) {
                ByteBufCodecs.STRING_UTF8.encode(buf, entry.getKey());
                ByteBufCodecs.STRING_UTF8.encode(buf, entry.getValue());
            }
        },
        buf -> {
            int id = ByteBufCodecs.INT.decode(buf);
            int size = ByteBufCodecs.INT.decode(buf);
            Map<String, String> map = new HashMap<>(size);
            for (int i = 0; i < size; i++) {
                String key = ByteBufCodecs.STRING_UTF8.decode(buf);
                String value = ByteBufCodecs.STRING_UTF8.decode(buf);
                map.put(key, value);
            }
            return new TrainScriptDataPayload(id, map);
        }
    );

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return TYPE;
    }

    public static void handleOnClient(TrainScriptDataPayload payload, IPayloadContext context) {
        context.enqueueWork(() -> {
            Minecraft minecraft = Minecraft.getInstance();
            if (minecraft.level == null) return;
            if (minecraft.level.getEntity(payload.trainEntityId()) instanceof TrainEntity train) {
                train.applyScriptDataSync(payload.data());
            }
        });
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/network/TrainScriptDataPayload.kt
package cc.mirukuneko.realtrainmodrenewed.network

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed
import cc.mirukuneko.realtrainmodrenewed.entity.TrainEntity
import io.netty.buffer.ByteBuf
import net.minecraft.client.Minecraft
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.Identifier
import net.neoforged.neoforge.network.handling.IPayloadContext
import java.util.HashMap

@JvmRecord
data class TrainScriptDataPayload(
    val trainEntityId: Int,
    val data: Map<String, String>,
) : CustomPacketPayload {
    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

    companion object {
        @JvmField
        val TYPE: CustomPacketPayload.Type<TrainScriptDataPayload> = CustomPacketPayload.Type(
            Identifier.fromNamespaceAndPath(RealTrainModRenewed.MODID, "train_script_data")
        )

        @JvmField
        val STREAM_CODEC: StreamCodec<ByteBuf, TrainScriptDataPayload> = StreamCodec.of(
            { buf, payload ->
                ByteBufCodecs.INT.encode(buf, payload.trainEntityId)
                ByteBufCodecs.INT.encode(buf, payload.data.size)
                for ((key, value) in payload.data) {
                    ByteBufCodecs.STRING_UTF8.encode(buf, key)
                    ByteBufCodecs.STRING_UTF8.encode(buf, value)
                }
            },
            { buf ->
                val id = ByteBufCodecs.INT.decode(buf)
                val size = ByteBufCodecs.INT.decode(buf)
                val map = HashMap<String, String>(size)
                repeat(size) {
                    val key = ByteBufCodecs.STRING_UTF8.decode(buf)
                    val value = ByteBufCodecs.STRING_UTF8.decode(buf)
                    map[key] = value
                }
                TrainScriptDataPayload(id, map)
            },
        )

        @JvmStatic
        fun handleOnClient(payload: TrainScriptDataPayload, context: IPayloadContext) {
            context.enqueueWork {
                val minecraft = Minecraft.getInstance()
                val level = minecraft.level ?: return@enqueueWork
                val train = level.getEntity(payload.trainEntityId) as? TrainEntity ?: return@enqueueWork
                train.applyScriptDataSync(payload.data)
            }
        }
    }
}
